File: solutions/python/07/closest_pair/main.py
def dist2(a, b):
    return (a[0] - b[0]) ** 2 + (a[1] - b[1]) ** 2


def closest_pair(points):
    points = sorted(points)
    n = len(points)
    if n <= 1:
        return 0

    def solve(l, r):
        if r - l <= 3:
            best = 10**30
            for i in range(l, r):
                for j in range(i + 1, r):
                    best = min(best, dist2(points[i], points[j]))
            return best
        mid = (l + r) // 2
        midx = points[mid][0]
        d = min(solve(l, mid), solve(mid, r))
        strip = [p for p in points[l:r] if (p[0] - midx) ** 2 < d]
        strip.sort(key=lambda p: p[1])
        for i in range(len(strip)):
            for j in range(i + 1, min(i + 8, len(strip))):
                d = min(d, dist2(strip[i], strip[j]))
        return d

    return solve(0, n)


def main() -> None:
    pts = [(0, 0), (1, 1), (3, 4), (0, 5), (2, 2)]
    print(closest_pair(pts) ** 0.5)


if __name__ == "__main__":
    main()

File: solutions/go/07/closest_pair/main.go
package main

import (
	"fmt"
	"math"
	"sort"
)

type point struct{ x, y int }

func dist2(a, b point) int {
	dx := a.x - b.x
	dy := a.y - b.y
	return dx*dx + dy*dy
}

func closestPair(points []point) int {
	sort.Slice(points, func(i, j int) bool { return points[i].x < points[j].x })
	n := len(points)
	if n <= 1 {
		return 0
	}
	var solve func(l, r int) int
	solve = func(l, r int) int {
		if r-l <= 3 {
			best := int(1e18)
			for i := l; i < r; i++ {
				for j := i + 1; j < r; j++ {
					best = min(best, dist2(points[i], points[j]))
				}
			}
			return best
		}
		mid := (l + r) / 2
		midx := points[mid].x
		d := min(solve(l, mid), solve(mid, r))
		var strip []point
		for _, p := range points[l:r] {
			dx := p.x - midx
			if dx*dx < d {
				strip = append(strip, p)
			}
		}
		sort.Slice(strip, func(i, j int) bool { return strip[i].y < strip[j].y })
		for i := 0; i < len(strip); i++ {
			for j := i + 1; j < len(strip) && j < i+8; j++ {
				d = min(d, dist2(strip[i], strip[j]))
			}
		}
		return d
	}
	return solve(0, n)
}

func main() {
	pts := []point{{0, 0}, {1, 1}, {3, 4}, {0, 5}, {2, 2}}
	fmt.Println(math.Sqrt(float64(closestPair(pts))))
}
